File: 14-TypeScript/Grey Multi-Tenant/packages/grey-php/src/Http/HttpClient.php
<?php

declare(strict_types=1);

namespace Grey\Sdk\Http;

use Grey\Sdk\Config\GreyOptions;
use Grey\Sdk\Error\GreyError;
use Grey\Sdk\Error\Result;
use GuzzleHttp\Client;
use GuzzleHttp\Exception\GuzzleException;
use GuzzleHttp\RequestOptions;

/**
 * HTTP client for making requests to the Grey API.
 */
final class HttpClient
{
    private Client $client;
    private ?string $accessToken = null;

    public function __construct(
        private readonly GreyOptions $options,
    ) {
        $this->client = new Client([
            'base_uri' => $options->getBaseUrl(),
            'timeout' => $options->getTimeoutSeconds(),
            'headers' => array_merge(
                ['Content-Type' => 'application/json', 'Accept' => 'application/json'],
                $options->getMetadata(),
            ),
        ]);
    }

    /**
     * Sets the access token for authenticated requests.
     */
    public function setAccessToken(?string $token): void
    {
        $this->accessToken = $token;
    }

    /**
     * Gets the current access token.
     */
    public function getAccessToken(): ?string
    {
        return $this->accessToken;
    }

    /**
     * Returns true if authenticated.
     */
    public function isAuthenticated(): bool
    {
        return $this->accessToken !== null;
    }

    /**
     * Makes a GET request.
     *
     * @return Result<array>
     */
    public function get(string $path, array $query = [], bool $requireAuth = true): Result
    {
        return $this->request('GET', $path, [
            RequestOptions::QUERY => $query,
        ], $requireAuth);
    }

    /**
     * Makes a POST request.
     *
     * @return Result<array>
     */
    public function post(string $path, array $body = [], bool $requireAuth = true): Result
    {
        return $this->request('POST', $path, [
            RequestOptions::JSON => $body,
        ], $requireAuth);
    }

    /**
     * Makes a PUT request.
     *
     * @return Result<array>
     */
    public function put(string $path, array $body = [], bool $requireAuth = true): Result
    {
        return $this->request('PUT', $path, [
            RequestOptions::JSON => $body,
        ], $requireAuth);
    }

    /**
     * Makes a PATCH request.
     *
     * @return Result<array>
     */
    public function patch(string $path, array $body = [], bool $requireAuth = true): Result
    {
        return $this->request('PATCH', $path, [
            RequestOptions::JSON => $body,
        ], $requireAuth);
    }

    /**
     * Makes a DELETE request.
     *
     * @return Result<array>
     */
    public function delete(string $path, array $body = [], bool $requireAuth = true): Result
    {
        $options = [];
        if (!empty($body)) {
            $options[RequestOptions::JSON] = $body;
        }
        return $this->request('DELETE', $path, $options, $requireAuth);
    }

    /**
     * Makes an HTTP request.
     *
     * @return Result<array>
     */
    private function request(string $method, string $path, array $options = [], bool $requireAuth = true): Result
    {
        if ($requireAuth && !$this->isAuthenticated()) {
            return Result::err(GreyError::unauthorized());
        }

        try {
            $headers = $options['headers'] ?? [];

            if ($this->accessToken !== null) {
                $headers['Authorization'] = 'Bearer ' . $this->accessToken;
            }

            $options['headers'] = $headers;

            $response = $this->client->request($method, $path, $options);
            $statusCode = $response->getStatusCode();
            $body = (string) $response->getBody();

            $data = [];
            if (!empty($body)) {
                $decoded = json_decode($body, true, 512, JSON_THROW_ON_ERROR);
                $data = is_array($decoded) ? $decoded : ['data' => $decoded];
            }

            if ($statusCode >= 200 && $statusCode < 300) {
                return Result::ok($data);
            }

            return Result::err(GreyError::fromHttpResponse($statusCode, $data));
        } catch (GuzzleException $e) {
            return Result::err(GreyError::fromException($e));
        } catch (\JsonException $e) {
            return Result::err(GreyError::validation('Invalid JSON response: ' . $e->getMessage()));
        } catch (\Throwable $e) {
            return Result::err(GreyError::fromException($e));
        }
    }
}

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-ruby/lib/grey_sdk/http/http_client.rb
# frozen_string_literal: true

require "net/http"
require "uri"
require "json"

module GreySdk
  module Http
    # Shared HTTP client for making requests to the Grey API
    class HttpClient
      attr_reader :options, :access_token

      # @param options [GreySdk::Config::Options] Client options
      def initialize(options)
        @options = options
        @access_token = nil
      end

      # Sets the access token for authenticated requests
      # @param token [String, nil] Access token
      def access_token=(token)
        @access_token = token
      end

      # Makes a GET request
      # @param path [String] Request path
      # @param params [Hash] Query parameters
      # @return [GreySdk::Error::Result]
      def get(path, params = {})
        uri = build_uri(path, params)
        request = Net::HTTP::Get.new(uri)
        execute(uri, request)
      end

      # Makes a POST request
      # @param path [String] Request path
      # @param body [Hash] Request body
      # @return [GreySdk::Error::Result]
      def post(path, body = {})
        uri = build_uri(path)
        request = Net::HTTP::Post.new(uri)
        request.body = body.to_json
        execute(uri, request)
      end

      # Makes a PUT request
      # @param path [String] Request path
      # @param body [Hash] Request body
      # @return [GreySdk::Error::Result]
      def put(path, body = {})
        uri = build_uri(path)
        request = Net::HTTP::Put.new(uri)
        request.body = body.to_json
        execute(uri, request)
      end

      # Makes a PATCH request
      # @param path [String] Request path
      # @param body [Hash] Request body
      # @return [GreySdk::Error::Result]
      def patch(path, body = {})
        uri = build_uri(path)
        request = Net::HTTP::Patch.new(uri)
        request.body = body.to_json
        execute(uri, request)
      end

      # Makes a DELETE request
      # @param path [String] Request path
      # @param params [Hash] Query parameters
      # @return [GreySdk::Error::Result]
      def delete(path, params = {})
        uri = build_uri(path, params)
        request = Net::HTTP::Delete.new(uri)
        execute(uri, request)
      end

      private

      # Builds a full URI from path and optional query parameters
      # @param path [String] Request path
      # @param params [Hash] Query parameters
      # @return [URI]
      def build_uri(path, params = {})
        uri = URI.join(@options.base_url, path)
        uri.query = URI.encode_www_form(params) unless params.empty?
        uri
      end

      # Applies default headers to a request
      # @param request [Net::HTTPRequest] The request
      def apply_headers(request)
        request["Content-Type"] = "application/json"
        request["Accept"] = "application/json"

        @options.headers.each do |key, value|
          request[key.to_s] = value
        end

        request["Authorization"] = "Bearer #{@access_token}" if @access_token
      end

      # Creates an HTTP connection with proper settings
      # @param uri [URI] Request URI
      # @return [Net::HTTP]
      def create_connection(uri)
        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = @options.use_tls
        http.open_timeout = @options.timeout
        http.read_timeout = @options.timeout
        http.verify_mode = OpenSSL::SSL::VERIFY_PEER if @options.use_tls
        http
      end

      # Executes an HTTP request and returns a Result
      # @param uri [URI] Request URI
      # @param request [Net::HTTPRequest] The request
      # @return [GreySdk::Error::Result]
      def execute(uri, request)
        apply_headers(request)
        http = create_connection(uri)

        response = http.request(request)
        handle_response(response)
      rescue Errno::ECONNREFUSED, Errno::ECONNRESET, Errno::EHOSTUNREACH => e
        Error::Result.err(Error::GreyError.network(e.message))
      rescue Net::OpenTimeout, Net::ReadTimeout => e
        Error::Result.err(Error::GreyError.timeout(e.message))
      rescue JSON::ParserError => e
        Error::Result.err(Error::GreyError.validation("Invalid JSON: #{e.message}"))
      rescue StandardError => e
        Error::Result.err(Error::GreyError.unknown(e.message))
      end

      # Handles an HTTP response and returns a Result
      # @param response [Net::HTTPResponse] The response
      # @return [GreySdk::Error::Result]
      def handle_response(response)
        body = parse_body(response.body)
        status = response.code.to_i

        if status >= 200 && status < 300
          Error::Result.ok(body)
        else
          error = Error::GreyError.from_http_response(status, body)
          Error::Result.err(error)
        end
      end

      # Parses a response body as JSON
      # @param body [String, nil] Response body
      # @return [Hash, nil]
      def parse_body(body)
        return nil if body.nil? || body.empty?

        JSON.parse(body)
      rescue JSON::ParserError
        { "raw" => body }
      end
    end
  end
end
